src/JsonApiSerializer/JsonConverters/ResourceObjectConverter.cs

```diff
@@ -149,146 +167,238 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
+                }
+
+                // at this point we must be in an attributes block as the property is neither well known nor 
+                // a relationship. if we haven't written out the attributes block yet, now is the time
+                if (!didWriteAttributes)
+                {
+                    didWriteAttributes = true;
+
+                    writer.WritePropertyName(PropertyNames.Attributes, false);
+                    writer.WriteStartObject();
+                }
+
+                writer.WritePropertyName(property.PropertyName, false);
+
+                if (property.MemberConverter != null && property.MemberConverter.CanWrite)
+                {
+                    property.MemberConverter.WriteJson(writer, propertyValue, serializer);
+                }
+                else
+                {
+                    // writing a property via `serializer.Serialize` is slow, 
+                    // for certain well known primitive types we can write the 
+                    // value directly to the JsonWriter
+
+                    if (propertyValue is string s)
+                    {
+                        writer.WriteValue(s);
+                    }
+                    else if (propertyValue is int i)
+                    {
+                        writer.WriteValue(i);
+                    }
+                    else
+                    {
+                        serializer.Serialize(writer, propertyValue);
+                    }
+                }
+            }
+
+            // if we did write attributes make sure to close out the attributes block
+            if (didWriteAttributes)
+            {
                 writer.WriteEndObject();
             }
 
-            //output our relationships
-            if (relationships.Count > 0)
+            // output our relationships
+            if (relationships != null)
             {
-                writer.WritePropertyName(PropertyNames.Relationships);
+                writer.WritePropertyName(PropertyNames.Relationships, false);
                 writer.WriteStartObject();
                 foreach (var relationship in relationships)
                 {
-                    writer.WritePropertyName(relationship.Key.PropertyName);
+                    writer.WritePropertyName(relationship.Key, false);
+
                     serializer.Serialize(writer, relationship.Value);
                 }
                 writer.WriteEndObject();
             }
 
             writer.WriteEndObject();
 
-            //add reference to this type, so others can reference it
+            if (id == null)
+            {
+                return;
+            }
+
+            // add reference to this type, so others can reference it
             var serializationData = SerializationData.GetSerializationData(writer);
-            var reference = new ResourceObjectReference(id?.ToString(), type?.ToString());
-            serializationData.Included[reference] = value;

```

Comment:
I feel like excluding this will break something. Might need to test some circular references between data and included elements

Reply:
This is the most important thing I’d like to change. The include tracking is probably the most resource intensive part of this whole process - especially given the implementation of the OrderedDictionary. 

I think for reads you are definitely right, but for writes I checked all the references and I don’t think this will cause a problem. The included items are always filtered by rendered items. 

I have some ideas for how to make this better, but would rather do it as part of another PR. If there are additional tests you think are worth adding on top of the existing ones to double check circular references I can add those.